### ISSUE TITLE
SetDebuggerOptions fails under VS 2017 version 15.7.1

### ISSUE THREAD

**Issue by govert** [Excel-DNA/ExcelDna#181]
@caioproiete I still have to take a closer look, but it seems the NuGet package's SetDebuggerOptions action fails after I upgraded to VS 2017 version 15.7.1.

`1>"c:\Temp\CsTest157\CsTest157.csproj" (Rebuild target) (1) ->
1>(ExcelDnaSetDebuggerOptions target) -> 
1>  MSBUILD : ExcelDnaSetDebuggerOptions warning DNA277884429: Unable to set the debugger options within Visual Studio.
1>`

I have not yet tried to debug it or reproduce with a clean machine.

**Comment by caioproiete** [comment]
@govert Are you able to reproduce it on the same machine after restarting Visual Studio yourself (and not the installer that performed the upgrade) and/or after a reboot of the same machine?

I'm guessing this was caused by a problem with the VS upgrade itself, as I was not able to reproduce it on my VM (`ExcelDna.AddIn v0.35.1-beta2`):

![vs](https://user-images.githubusercontent.com/177608/39970094-0f64a3b4-56bc-11e8-9018-4f9df6cdb4a3.png)

**Comment by govert** [comment]
@caioproiete You fixed it!

I closed all my VS instances and restarted and now it seems fine.

Do you have any speculation on why this might be?
How did you know to restart VS again?

**Comment by caioproiete** [comment]
Great!

It was just a hunch. VS2017 hasn't been stable lately and I've had issues before with packages that rely on custom MSBuild tasks, such as [Fody](https://github.com/Fody/Fody) for example, which I was only able to fix by restarting VS every now and then.

I'm guessing there are deadlocks happening inside VS sometimes that block access to its object model. The warning you're seeing in the Excel-DNA build happens when we get a "Visual Studio is busy" exception when trying to access its object model (i.e. DTE).

I don't think there's much we can do on our side, other than asking the user to restart VS when that happens.

**Comment by govert** [comment]
@caioproiete Sorry - the restart worked for the small single project, but on the slightly simple three-project solution, it still gives the error after restart of Windows and VS.

**Comment by govert** [comment]
The problem (where restart does not fix) has to do with solution folders.

@caioproiete Not urgent and I can try fix too, but I'd be happy to know whether you can reproduce the problem:

* Open Visual Studio
* Create a new 'Blank Solution' in C:\Test called "TheTestSolution"
* In the Solution Explorer, add a new Solution Folder called "TheFolder"
* Under "TheFolder" add a new Project (Class Library (.NET Framework) called "TheLibrary"
* `Install-Package -Pre ExcelDna.AddIn` (note that it shows the default project in the Package Manager Console as "TheFolder\TheLibrary"
* Press F5 to build and start debugging - fails with the message here.

Restart VS, rebuild still doesn't let the debugger be set.

Build diagnostics:
```
1>Target "ExcelDnaSetDebuggerOptions: (TargetId:16)" in file "C:\Test\TheTestSolution\packages\ExcelDna.AddIn.0.35.1-beta2\build\ExcelDna.AddIn.targets" from project "C:\Test\TheTestSolution\TheLibrary\TheLibrary.csproj" (target "Build" depends on it):
1>Task "Message" (TaskId:17)
1>  Task Parameter:Text=--- (TaskId:17)
1>  Task Parameter:Importance=High (TaskId:17)
1>  --- (TaskId:17)
1>Done executing task "Message". (TaskId:17)
1>Task "Error" skipped, due to false condition; ('$(ExcelDna32BitAddInSuffix)' == '$(ExcelDna64BitAddInSuffix)') was evaluated as ('' == '64').
1>Added Item(s): 
1>    ExcelDnaFilesInProject=
1>        TheLibrary-AddIn.dna
1>        packages.config
1>        Properties\ExcelDna.Build.props
1>Using "SetDebuggerOptions" task from assembly "C:\Test\TheTestSolution\packages\ExcelDna.AddIn.0.35.1-beta2\build\..\tools\ExcelDna.AddIn.Tasks.dll".
1>Task "SetDebuggerOptions" (TaskId:18)
1>  Task Parameter:ProjectName=TheLibrary (TaskId:18)
1>  Task Parameter:
1>      FilesInProject=
1>          TheLibrary-AddIn.dna
1>          packages.config
1>          Properties\ExcelDna.Build.props
1>          TheLibrary-AddIn.dna
1>          packages.config
1>          Properties\ExcelDna.Build.props (TaskId:18)
1>  Task Parameter:OutDirectory=bin\Debug\ (TaskId:18)
1>  Task Parameter:FileSuffix64Bit=64 (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: Running SetDebuggerOptions MSBuild Task (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: Number of files in project: 6 (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: EXCEL.EXE path for debugging: C:\Program Files (x86)\Microsoft Office\root\Office16\EXCEL.EXE (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: Add-In for debugging: bin\Debug\TheLibrary-AddIn.xll (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: ----Arguments---- (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: ProjectName: TheLibrary (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: ExcelExePath:  (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: AddInForDebugging:  (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: FilesInProject: 6 (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: OutDirectory: bin\Debug\ (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: FileSuffix32Bit:  (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: FileSuffix64Bit: 64 (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: ----------------- (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: Start GetProjectByName (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: VS Process ID: 11600 (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: DevToolsEnvironment?: System.__ComObject (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: Did not find project (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: Looked for TheLibrary (TaskId:18)
1>  ExcelDnaSetDebuggerOptions: List of projects: (TaskId:18)
1>  ExcelDnaSetDebuggerOptions:     __ComObject  -  Is EnvDTE? True  -  TheFolder (TaskId:18)
1>MSBUILD : ExcelDnaSetDebuggerOptions warning DNA277884429: Unable to set the debugger options within Visual Studio.
1>Done executing task "SetDebuggerOptions". (TaskId:18)
```

**Comment by caioproiete** [comment]
Nice! I was able to reproduce it with the steps you outlined

> * Open Visual Studio
> * Create a new 'Blank Solution' in C:\Test called "TheTestSolution"
> * In the Solution Explorer, add a new Solution Folder called "TheFolder"
> * Under "TheFolder" add a new Project (Class Library (.NET Framework) called "TheLibrary"
> * Install-Package -Pre ExcelDna.AddIn (note that it shows the default project in the Package Manager Console > as * "TheFolder\TheLibrary"
> * Press F5 to build and start debugging - fails with the message here.

I repeated the process without creating the solution folder, and the problem goes away, so the solution folder seems to be the issue here.

I'll investigate and send PR.

**Comment by caioproiete** [comment]
Good catch. The DTE doesn't return a flat list of projects, so we need to [check for solution folders and read its sub-projects](https://github.com/Excel-DNA/ExcelDna/commit/2b0b1ad7d5643574852bce921059f6a4e92f0e77#diff-ba2d9e47e4079b33a6ca8c914b09f241R162).

**Comment by caioproiete** [comment]
Fixed by #184
